src/V3Fork.cpp

```diff
@@ -0,0 +1,235 @@
+            taskp
+                = makeTask(beginp->fileline(), beginp->stmtsp()->unlinkFrBackWithNext(), taskName);
+            beginp->unlinkFrBack(&handle);
+            VL_DO_DANGLING(beginp->deleteTree(), beginp);
+        } else if (AstNodeStmt* stmtp = VN_CAST(nodep, NodeStmt)) {
+            const std::string taskName = generateTaskName(stmtp, "__FORK_STMT_");
+            taskp = makeTask(stmtp->fileline(), stmtp->unlinkFrBack(&handle), taskName);
+        } else if (AstFork* forkp = VN_CAST(nodep, Fork)) {
+            const std::string taskName = generateTaskName(forkp, "__FORK_NESTED_");
+            taskp = makeTask(forkp->fileline(), forkp->unlinkFrBack(&handle), taskName);
+        }
+
+        m_modp->addStmtsp(taskp);
+
+        AstTaskRef* const taskrefp
+            = new AstTaskRef{nodep->fileline(), taskp->name(), m_capturedVarRefsp};
+        AstStmtExpr* const taskcallp = new AstStmtExpr{nodep->fileline(), taskrefp};
+        // Replaced nodes will be revisited, so we don't need to "lift" the arguments
+        // as captures in case of nested forks.
+        handle.relink(taskcallp);
+        taskcallp->user1(true);
+    }
+
+    // VISITORS
+    void visit(AstFork* nodep) override {
+        bool nested = m_newProcess;
+
+        VL_RESTORER(m_forkLocalsp);
+        VL_RESTORER(m_newProcess);
+        VL_RESTORER(m_forkDepth)
+        if (!nodep->joinType().join()) {
+            ++m_forkDepth;
+            m_newProcess = true;
+            m_forkLocalsp.clear();
+            // Nested forks get moved into separate tasks
+            if (nested) {
+                visitTaskifiable(nodep);
+                return;
+            }
+        } else {
+            m_newProcess = false;
+        }
+        iterateChildren(nodep);
+    }
+    void visit(AstBegin* nodep) override { visitTaskifiable(nodep); }
+    void visit(AstNodeStmt* nodep) override { visitTaskifiable(nodep); }
+    void visit(AstVar* nodep) override {
+        if (m_forkDepth) m_forkLocalsp.insert(nodep);
+    }
+    void visit(AstVarRef* nodep) override {
+
+        // VL_KEEP_THIS ensures that we hold a handle to the class
+        if (m_forkDepth && !nodep->varp()->isFuncLocal() && nodep->varp()->isClassMember()) return;
+
+        if (m_forkDepth && (m_forkLocalsp.count(nodep->varp()) == 0)
+            && !nodep->varp()->lifetime().isStatic()) {
+            if (nodep->access().isWriteOrRW()) {
+                nodep->v3warn(E_TASKNSVAR,
+                              "Invalid reference: Process might outlive variable `"
+                                  << nodep->varp()->name()
+                                  << "`. Use it as read-only to initialize a "
+                                     "local copy at the beginning of the process, or "
+                                     "declare it as static. It is also possible to "
+                                     "refer by reference to `this` object's members.");
```

Comment:
TASKNSVAR doesn't seem to be right here given what the documentation says it is for.  Maybe UNSUPPORTED, or maybe a new warning?

```suggestion
                nodep->v3warn(E_TASKNSVAR,
                              "Invalid reference: Process might outlive variable `"
                                  << nodep->varp()->name()
                                  << "`.\n"
                                  << nodep->varp()->warnMore() << "... Suggest use it as read-only to initialize a "
                                     "local copy at the beginning of the process, or "
                                     "declare it as static. It is also possible to "
                                     "refer by reference to 'this' object's members.");
```

Reply:
Alright, I improved this message.

I made a new error code - `E_LIFETIME` for this one. I also added some documentation for it with examples of a cause and possible fixes.